### PR TITLE
fetch_support_indexeddb

### PR DESCRIPTION
**Pull request by juj**
Add FETCH_SUPPORT_INDEXEDDB option to allow disabling IndexedDB support for Fetch API, when not desired. Closure is not able to DCE away IndexedDB use, so need to manually allow the option.

### CODE REVIEW

**Comment by kripken** [comment]
@jgravelle-google, I think you might know this code a little?

If there's no one to review it, I don't see any issues at a high level with landing it.

**Comment by juj** [comment]
This was updated a little bit to help preprocess fetch worker with {{{ cDefines }}}. Do you want to take another look before merging?

File: src/parseTools.js [hunk 1]
```diff
@@ -1530,6 +1537,12 @@ function makeRetainedCompilerSettings() {
   return ret;
 }
 
+// Safe way to access a C define. We check that we don't add library functions with missing defines.
+function cDefine(key) {
```
Comment (kripken) [inline src/parseTools.js]: `cDefine` already exists in modules.js?

File: src/parseTools.js [hunk 2]
```diff
@@ -8,6 +8,13 @@
 
 //"use strict";
 
+// Load struct and define information.
+if (STRUCT_INFO) {
+  var temp = JSON.parse(read(STRUCT_INFO));
+  C_STRUCTS = temp.structs;
```
Comment (kripken) [inline src/parseTools.js]: this is done in compiler.js?

**Comment by kripken** [comment]
Oh, reading the previous comment, maybe they were intentional? If so, why the duplication?

**Comment by juj** [comment]
The code was duplicated so that parseTools.js would also load in `C_STRUCTS`. It does not see the contents of that without having that code in parseTools.js. I first moved `C_STRUCTS` handling fully to parseTool.js, but then other parts of the compiler got broken. So I did not know how to implement it without duplicating the code in both files. I think I'm not familiar enough with the interplay/code flow of these files, so just did the simple copy and paste thing. If there's a simpler way that doesn't duplicate, that would definitely be better

**Comment by kripken** [comment]
compiler.js runs first. Then it loads the others here:

https://github.com/emscripten-core/emscripten/blob/incoming/src/compiler.js#L195

modules is loaded first, so parseTools should have access to what it creates. To confirm, I added `printErr(JSON.stringify(C_STRUCTS));` at the very end of parseTools.js and it prints out ok as expected. So I think this should already work fine - does that command not work for you? And how can I reproduce the problem you are seeing?

**Comment by juj** [comment]
Hmm, that is surprising. The issue appeared for me when I wanted to add use of C_STRUCTS to Fetch.js (https://github.com/emscripten-core/emscripten/pull/7825/files#diff-8e44b1ae605757a332e1a084ddf7ac15R196) to get rid of the runtime offset fields (https://github.com/emscripten-core/emscripten/pull/7825/files#diff-8e44b1ae605757a332e1a084ddf7ac15L7). Without the changes to parseTools.js, any test using Fetch API would fail, e.g. `browser.test_fetch*`, saying that they are missing information about the fetch_attr_t and fetch_t field offsets.

I wonder then if it could be an initialization order issue, or similar?

**Comment by kripken** [comment]
Hmm, Fetch.js is a little special it seems, I see it's loaded from library-fetch.js and fetch-worker.js. The former seems fine - just normal library code. But the fetch-worker.js case is loaded from somewhere else I guess? That somewhere might be the preprocessor that runs later, that is, it does not run in the compiler.js invocation normally? If that's it, then we need to move the shared code to a location both compiler.js and the other code path use equally.

**Comment by juj** [comment]
Yeah, it's loaded from both to share common code. fetch-worker.js case is read from make_fetch_worker when it's preprocessing the source to produce the worker script. It runs the JS side preprocessor script that you added not that long ago.

**Comment by kripken** [comment]
Hmm, I commented out your parseTools changes and then ran `./tests/runner.py browser.test_fetch_to_memory` successfully.

**Comment by kripken** [comment]
On the other hand, I see that `tools/preprocessor.js` loads parseTools, but not modules. Still not sure why it fails for you and not me, but maybe try loading modules there too (before parseTools).

**Comment by juj** [comment]
> any test using Fetch API would fail, e.g. `browser.test_fetch*`

> `./tests/runner.py browser.test_fetch_to_memory` successfully.

Sorry, not any test using Fetch API, but any test using fetch worker. That test is a singlethreaded test, it does not run make_fetch_worker. Running `browser.test_fetch*` gives

```
======================================================================
ERROR: test_fetch_idb_delete (test_browser.browser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/clb/emsdk/emscripten/incoming/tests/test_browser.py", line 116, in decorated
    return f(self)
  File "/Users/clb/emsdk/emscripten/incoming/tests/test_browser.py", line 4173, in test_fetch_idb_delete
    self.btest('fetch/idb_delete.cpp', expected='0', args=['-s', 'USE_PTHREADS=1', '-s', 'FETCH_DEBUG=1', '-s', 'FETCH=1', '-s', 'WASM=0', '-s', 'PROXY_TO_PTHREAD=1'])
  File "/Users/clb/emsdk/emscripten/incoming/tests/runner.py", line 1308, in btest
    run_process(all_args)
  File "/Users/clb/emsdk/emscripten/incoming/tools/shared.py", line 170, in run_process
    result.check_returncode()
  File "/Users/clb/emsdk/emscripten/incoming/tools/shared.py", line 150, in check_returncode
    raise Py2CalledProcessError(returncode=self.returncode, cmd=self.args, output=self.stdout, stderr=self.stderr)
Py2CalledProcessError: Command '['/usr/bin/python', '/Users/clb/emsdk/emscripten/incoming/emcc.py', '-s', 'IN_TEST_HARNESS=1', '/Users/clb/emsdk/emscripten/incoming/tests/fetch/idb_delete.cpp', '-o', 'test.html', '-s', 'USE_PTHREADS=1', '-s', 'FETCH_DEBUG=1', '-s', 'FETCH=1', '-s', 'WASM=0', '-s', 'PROXY_TO_PTHREAD=1', '-DEMTEST_PORT_NUMBER=8888', '-include', '/Users/clb/emsdk/emscripten/incoming/tests/report_result.h']' returned non-zero exit status 1

======================================================================
ERROR: test_fetch_idb_store (test_browser.browser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/clb/emsdk/emscripten/incoming/tests/test_browser.py", line 116, in decorated
    return f(self)
  File "/Users/clb/emsdk/emscripten/incoming/tests/test_browser.py", line 4168, in test_fetch_idb_store
    self.btest('fetch/idb_store.cpp', expected='0', args=['-s', 'USE_PTHREADS=1', '-s', 'FETCH_DEBUG=1', '-s', 'FETCH=1', '-s', 'WASM=0', '-s', 'PROXY_TO_PTHREAD=1'])
  File "/Users/clb/emsdk/emscripten/incoming/tests/runner.py", line 1308, in btest
    run_process(all_args)
  File "/Users/clb/emsdk/emscripten/incoming/tools/shared.py", line 170, in run_process
    result.check_returncode()
  File "/Users/clb/emsdk/emscripten/incoming/tools/shared.py", line 150, in check_returncode
    raise Py2CalledProcessError(returncode=self.returncode, cmd=self.args, output=self.stdout, stderr=self.stderr)
Py2CalledProcessError: Command '['/usr/bin/python', '/Users/clb/emsdk/emscripten/incoming/emcc.py', '-s', 'IN_TEST_HARNESS=1', '/Users/clb/emsdk/emscripten/incoming/tests/fetch/idb_store.cpp', '-o', 'test.html', '-s', 'USE_PTHREADS=1', '-s', 'FETCH_DEBUG=1', '-s', 'FETCH=1', '-s', 'WASM=0', '-s', 'PROXY_TO_PTHREAD=1', '-DEMTEST_PORT_NUMBER=8888', '-include', '/Users/clb/emsdk/emscripten/incoming/tests/report_result.h']' returned non-zero exit status 1

======================================================================
ERROR: test_fetch_response_headers (test_browser.browser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/clb/emsdk/emscripten/incoming/tests/test_browser.py", line 116, in decorated
    return f(self)
  File "/Users/clb/emsdk/emscripten/incoming/tests/test_browser.py", line 4126, in test_fetch_response_headers
    self.btest('fetch/response_headers.cpp', expected='1', args=['--std=c++11', '-s', 'FETCH_DEBUG=1', '-s', 'FETCH=1', '-s', 'USE_PTHREADS=1', '-s', 'PROXY_TO_PTHREAD=1'], also_asmjs=True)
  File "/Users/clb/emsdk/emscripten/incoming/tests/runner.py", line 1319, in btest
    args + ['-s', 'WASM=0'], outfile, message, also_proxied=False, timeout=timeout)
  File "/Users/clb/emsdk/emscripten/incoming/tests/runner.py", line 1308, in btest
    run_process(all_args)
  File "/Users/clb/emsdk/emscripten/incoming/tools/shared.py", line 170, in run_process
    result.check_returncode()
  File "/Users/clb/emsdk/emscripten/incoming/tools/shared.py", line 150, in check_returncode
    raise Py2CalledProcessError(returncode=self.returncode, cmd=self.args, output=self.stdout, stderr=self.stderr)
Py2CalledProcessError: Command '['/usr/bin/python', '/Users/clb/emsdk/emscripten/incoming/emcc.py', '-s', 'IN_TEST_HARNESS=1', '/Users/clb/emsdk/emscripten/incoming/tests/fetch/response_headers.cpp', '-o', 'test.html', '--std=c++11', '-s', 'FETCH_DEBUG=1', '-s', 'FETCH=1', '-s', 'USE_PTHREADS=1', '-s', 'PROXY_TO_PTHREAD=1', '-DEMTEST_PORT_NUMBER=8888', '-include', '/Users/clb/emsdk/emscripten/incoming/tests/report_result.h', '-s', 'WASM=0', '-DEMTEST_PORT_NUMBER=8888', '-include', '/Users/clb/emsdk/emscripten/incoming/tests/report_result.h']' returned non-zero exit status 1

======================================================================
ERROR: test_fetch_sync_fetch_in_main_thread (test_browser.browser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/clb/emsdk/emscripten/incoming/tests/test_browser.py", line 116, in decorated
    return f(self)
  File "/Users/clb/emsdk/emscripten/incoming/tests/test_browser.py", line 4164, in test_fetch_sync_fetch_in_main_thread
    self.btest('fetch/sync_fetch_in_main_thread.cpp', expected='0', args=['--std=c++11', '-s', 'FETCH_DEBUG=1', '-s', 'FETCH=1', '-s', 'WASM=0', '-s', 'USE_PTHREADS=1', '-s', 'PROXY_TO_PTHREAD=1'])
  File "/Users/clb/emsdk/emscripten/incoming/tests/runner.py", line 1308, in btest
    run_process(all_args)
  File "/Users/clb/emsdk/emscripten/incoming/tools/shared.py", line 170, in run_process
    result.check_returncode()
  File "/Users/clb/emsdk/emscripten/incoming/tools/shared.py", line 150, in check_returncode
    raise Py2CalledProcessError(returncode=self.returncode, cmd=self.args, output=self.stdout, stderr=self.stderr)
Py2CalledProcessError: Command '['/usr/bin/python', '/Users/clb/emsdk/emscripten/incoming/emcc.py', '-s', 'IN_TEST_HARNESS=1', '/Users/clb/emsdk/emscripten/incoming/tests/fetch/sync_fetch_in_main_thread.cpp', '-o', 'test.html', '--std=c++11', '-s', 'FETCH_DEBUG=1', '-s', 'FETCH=1', '-s', 'WASM=0', '-s', 'USE_PTHREADS=1', '-s', 'PROXY_TO_PTHREAD=1', '-DEMTEST_PORT_NUMBER=8888', '-include', '/Users/clb/emsdk/emscripten/incoming/tests/report_result.h']' returned non-zero exit status 1

======================================================================
ERROR: test_fetch_sync_xhr (test_browser.browser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/clb/emsdk/emscripten/incoming/tests/test_browser.py", line 116, in decorated
    return f(self)
  File "/Users/clb/emsdk/emscripten/incoming/tests/test_browser.py", line 4149, in test_fetch_sync_xhr
    self.btest('fetch/sync_xhr.cpp', expected='1', args=['--std=c++11', '-s', 'FETCH_DEBUG=1', '-s', 'FETCH=1', '-s', 'WASM=0', '-s', 'USE_PTHREADS=1', '-s', 'PROXY_TO_PTHREAD=1'])
  File "/Users/clb/emsdk/emscripten/incoming/tests/runner.py", line 1308, in btest
    run_process(all_args)
  File "/Users/clb/emsdk/emscripten/incoming/tools/shared.py", line 170, in run_process
    result.check_returncode()
  File "/Users/clb/emsdk/emscripten/incoming/tools/shared.py", line 150, in check_returncode
    raise Py2CalledProcessError(returncode=self.returncode, cmd=self.args, output=self.stdout, stderr=self.stderr)
Py2CalledProcessError: Command '['/usr/bin/python', '/Users/clb/emsdk/emscripten/incoming/emcc.py', '-s', 'IN_TEST_HARNESS=1', '/Users/clb/emsdk/emscripten/incoming/tests/fetch/sync_xhr.cpp', '-o', 'test.html', '--std=c++11', '-s', 'FETCH_DEBUG=1', '-s', 'FETCH=1', '-s', 'WASM=0', '-s', 'USE_PTHREADS=1', '-s', 'PROXY_TO_PTHREAD=1', '-DEMTEST_PORT_NUMBER=8888', '-include', '/Users/clb/emsdk/emscripten/incoming/tests/report_result.h']' returned non-zero exit status 1

----------------------------------------------------------------------
Ran 10 tests in 195.712s

FAILED (errors=5)
```

**Comment by kripken** [comment]
Thanks, now I see the error. Fix is simple as I hoped, see #8091

**Comment by juj** [comment]
Thanks, much appreciated! Updated this PR to drop the custom changes that I had, looks like all passes. How does this look now?

**Pull request merged**